File: docker/health/health-monitor.py
#!/usr/bin/env python3
import sys
import json
import time
import argparse
from datetime import datetime
from typing import Dict, List, Optional, Callable, Any
from concurrent.futures import ThreadPoolExecutor, TimeoutError
import asyncio
import aiohttp

class HealthMonitor:
    def __init__(self, config: Dict[str, Any] = None):
        config = config or {}
        self.services = config.get('services', [])
        self.check_interval = config.get('checkInterval', 30000) / 1000  # Convert to seconds
        self.timeout = config.get('timeout', 5000) / 1000  # Convert to seconds
        self.health_checks = {}
        self.service_status = {}
        self.start_time = time.time()
        self.initialized = set()
        
        # Initialize service status
        for service in self.services:
            self.service_status[service] = {
                'status': 'pending',
                'lastCheck': None,
                'message': 'Not checked yet'
            }
    
    async def check_health(self, service_name: Optional[str] = None) -> Dict:
        if not service_name:
            # Check overall system health
            tasks = [self.check_service_health(service) for service in self.services]
            all_statuses = await asyncio.gather(*tasks)
            
            unhealthy = sum(1 for s in all_statuses if s['status'] == 'unhealthy')
            degraded = sum(1 for s in all_statuses if s['status'] == 'degraded')
            
            overall_status = 'healthy'
            if unhealthy > 0:
                overall_status = 'unhealthy'
            elif degraded > 0:
                overall_status = 'degraded'
            
            return {
                'status': overall_status,
                'details': {
                    'services': {s['service']: s for s in all_statuses},
                    'timestamp': datetime.utcnow().isoformat(),
                    'uptime': time.time() - self.start_time
                },
                'timestamp': datetime.utcnow().isoformat()
            }
        
        # Check specific service
        return await self.check_service_health(service_name)
    
    async def check_service_health(self, service_name: str) -> Dict:
        check_fn = self.health_checks.get(service_name)
        
        if not check_fn:
            # No custom check registered, use default
            status = self.service_status.get(service_name, {})
            return {
                'service': service_name,
                'status': status.get('status', 'unknown'),
                'details': status,
                'timestamp': datetime.utcnow().isoformat()
            }
        
        try:
            # Run health check with timeout
            result = await asyncio.wait_for(
                check_fn(),
                timeout=self.timeout
            )
            
            status = 'healthy' if result.get('healthy') else 'unhealthy'
            self.service_status[service_name] = {
                'status': status,
                'lastCheck': datetime.utcnow().isoformat(),
                'message': result.get('message', 'OK')
            }
            
            return {
                'service': service_name,
                'status': status,
                'details': result,
                'timestamp': datetime.utcnow().isoformat()
            }
        except (TimeoutError, asyncio.TimeoutError):
            self.service_status[service_name] = {
                'status': 'unhealthy',
                'lastCheck': datetime.utcnow().isoformat(),
                'message': 'Health check timeout'
            }
            
            return {
                'service': service_name,
                'status': 'unhealthy',
                'details': {'error': 'Health check timeout'},
                'timestamp': datetime.utcnow().isoformat()
            }
        except Exception as e:
            self.service_status[service_name] = {
                'status': 'unhealthy',
                'lastCheck': datetime.utcnow().isoformat(),
                'message': str(e)
            }
            
            return {
                'service': service_name,
                'status': 'unhealthy',
                'details': {'error': str(e)},
                'timestamp': datetime.utcnow().isoformat()
            }
    
    async def liveness_probe(self) -> Dict:
        """Simple liveness check - is the process alive?"""
        return {
            'alive': True,
            'timestamp': datetime.utcnow().isoformat()
        }
    
    async def readiness_probe(self) -> Dict:
        """Check if all required services are ready"""
        health = await self.check_health()
        ready = health['status'] != 'unhealthy'
        
        return {
            'ready': ready,
            'services': health['details']['services'],
            'timestamp': datetime.utcnow().isoformat()
        }
    
    async def startup_probe(self) -> Dict:
        """Check which services have been initialized"""
        initialized = list(self.initialized)
        pending = [s for s in self.services if s not in self.initialized]
        
        return {
            'started': len(pending) == 0,
            'initialized': initialized,
            'pending': pending
        }

def main():
    parser = argparse.ArgumentParser(description='Health check monitor')
    parser.add_argument('--check', choices=['liveness', 'readiness', 'startup'],
                       help='Type of health check to perform')
    args = parser.parse_args()
    
    if args.check:
        monitor = HealthMonitor({'services': []})
        
        async def run_check():
            if args.check == 'liveness':
                await monitor.liveness_probe()
                return 0
            elif args.check == 'readiness':
                probe = await monitor.readiness_probe()
                return 0 if probe['ready'] else 1
            elif args.check == 'startup':
                probe = await monitor.startup_probe()
                return 0 if probe['started'] else 1
        
        exit_code = asyncio.run(run_check())
        sys.exit(exit_code)

if __name__ == '__main__':
    main()

File: docker/health/health-monitor.js
const express = require('express');
const http = require('http');

class HealthMonitor {
  constructor(config = {}) {
    this.services = config.services || [];
    this.checkInterval = config.checkInterval || 30000;
    this.timeout = config.timeout || 5000;
    this.healthChecks = new Map();
    this.serviceStatus = new Map();
    this.startTime = Date.now();
    this.initialized = new Set();
    
    // Initialize service status
    this.services.forEach(service => {
      this.serviceStatus.set(service, {
        status: 'pending',
        lastCheck: null,
        message: 'Not checked yet'
      });
    });
    
    // Start periodic health checks
    this.startPeriodicChecks();
  }

  async checkHealth(serviceName) {
    if (!serviceName) {
      // Check overall system health
      const allStatuses = await Promise.all(
        this.services.map(service => this.checkServiceHealth(service))
      );
      
      const unhealthy = allStatuses.filter(s => s.status === 'unhealthy').length;
      const degraded = allStatuses.filter(s => s.status === 'degraded').length;
      
      let overallStatus = 'healthy';
      if (unhealthy > 0) overallStatus = 'unhealthy';
      else if (degraded > 0) overallStatus = 'degraded';
      
      return {
        status: overallStatus,
        details: {
          services: Object.fromEntries(
            allStatuses.map(s => [s.service, s])
          ),
          timestamp: new Date(),
          uptime: Date.now() - this.startTime
        },
        timestamp: new Date()
      };
    }
    
    // Check specific service
    return await this.checkServiceHealth(serviceName);
  }

  async checkServiceHealth(serviceName) {
    const checkFn = this.healthChecks.get(serviceName);
    
    if (!checkFn) {
      // No custom check registered, use default
      return {
        service: serviceName,
        status: this.serviceStatus.get(serviceName)?.status || 'unknown',
        details: this.serviceStatus.get(serviceName) || {},
        timestamp: new Date()
      };
    }
    
    try {
      const result = await Promise.race([
        checkFn(),
        new Promise((_, reject) => 
          setTimeout(() => reject(new Error('Health check timeout')), this.timeout)
        )
      ]);
      
      const status = result.healthy ? 'healthy' : 'unhealthy';
      this.serviceStatus.set(serviceName, {
        status,
        lastCheck: new Date(),
        message: result.message || 'OK'
      });
      
      return {
        service: serviceName,
        status,
        details: result,
        timestamp: new Date()
      };
    } catch (error) {
      this.serviceStatus.set(serviceName, {
        status: 'unhealthy',
        lastCheck: new Date(),
        message: error.message
      });
      
      return {
        service: serviceName,
        status: 'unhealthy',
        details: { error: error.message },
        timestamp: new Date()
      };
    }
  }

  async registerHealthCheck(serviceName, checkFn) {
    if (typeof checkFn !== 'function') {
      throw new Error('Health check must be a function');
    }
    
    this.healthChecks.set(serviceName, checkFn);
    
    // Add to services if not already present
    if (!this.services.includes(serviceName)) {
      this.services.push(serviceName);
      this.serviceStatus.set(serviceName, {
        status: 'pending',
        lastCheck: null,
        message: 'Not checked yet'
      });
    }
    
    // Run initial check
    await this.checkServiceHealth(serviceName);
    
    return { registered: true };
  }

  async livenessProbe() {
    // Simple liveness check - is the process alive?
    return {
      alive: true,
      timestamp: new Date()
    };
  }

  async readinessProbe() {
    // Check if all required services are ready
    const health = await this.checkHealth();
    const ready = health.status !== 'unhealthy';
    
    return {
      ready,
      services: health.details.services,
      timestamp: new Date()
    };
  }

  async startupProbe() {
    // Check which services have been initialized
    const initialized = Array.from(this.initialized);
    const pending = this.services.filter(s => !this.initialized.has(s));
    
    return {
      started: pending.length === 0,
      initialized,
      pending
    };
  }

  createHealthEndpoint(options = {}) {
    const router = express.Router();
    const { path = '/health', detailed = false, auth = false } = options;
    
    // Main health endpoint
    router.get(path, async (req, res) => {
      try {
        const health = await this.checkHealth();
        const statusCode = health.status === 'healthy' ? 200 : 
                          health.status === 'degraded' ? 200 : 503;
        
        if (detailed) {
          res.status(statusCode).json(health);
        } else {
          res.status(statusCode).json({
            status: health.status,
            timestamp: health.timestamp
          });
        }
      } catch (error) {
        res.status(500).json({
          status: 'error',
          error: error.message
        });
      }
    });
    
    // Liveness probe
    router.get(`${path}/live`, async (req, res) => {
      const probe = await this.livenessProbe();
      res.status(200).json(probe);
    });
    
    // Readiness probe
    router.get(`${path}/ready`, async (req, res) => {
      const probe = await this.readinessProbe();
      res.status(probe.ready ? 200 : 503).json(probe);
    });
    
    // Startup probe
    router.get(`${path}/startup`, async (req, res) => {
      const probe = await this.startupProbe();
      res.status(probe.started ? 200 : 503).json(probe);
    });
    
    return router;
  }

  createMetricsEndpoint(options = {}) {
    const router = express.Router();
    const { path = '/metrics', format = 'json' } = options;
    
    router.get(path, async (req, res) => {
      const health = await this.checkHealth();
      const metrics = {
        uptime: Date.now() - this.startTime,
        services: this.services.length,
        healthy: this.services.filter(s => 
          this.serviceStatus.get(s)?.status === 'healthy'
        ).length,
        unhealthy: this.services.filter(s => 
          this.serviceStatus.get(s)?.status === 'unhealthy'
        ).length,
        checks_total: this.healthChecks.size
      };
      
      if (format === 'prometheus') {
        // Convert to Prometheus format
        let output = '';
        output += `# HELP health_uptime_seconds Service uptime in seconds\n`;
        output += `# TYPE health_uptime_seconds gauge\n`;
        output += `health_uptime_seconds ${metrics.uptime / 1000}\n\n`;
        
        output += `# HELP health_services_total Total number of services\n`;
        output += `# TYPE health_services_total gauge\n`;
        output += `health_services_total ${metrics.services}\n\n`;
        
        output += `# HELP health_services_healthy Number of healthy services\n`;
        output += `# TYPE health_services_healthy gauge\n`;
        output += `health_services_healthy ${metrics.healthy}\n\n`;
        
        output += `# HELP health_services_unhealthy Number of unhealthy services\n`;
        output += `# TYPE health_services_unhealthy gauge\n`;
        output += `health_services_unhealthy ${metrics.unhealthy}\n`;
        
        res.set('Content-Type', 'text/plain; version=0.0.4');
        res.send(output);
      } else {
        res.json(metrics);
      }
    });
    
    return router;
  }

  async checkDependencies() {
    const dependencies = {
      'security-api': false,
      'rate-limiter': false,
      ...Object.fromEntries(this.services.map(s => [s, false]))
    };
    
    // Check each dependency
    for (const [dep, _] of Object.entries(dependencies)) {
      const health = await this.checkServiceHealth(dep);
      dependencies[dep] = health.status === 'healthy';
    }
    
    const missing = Object.entries(dependencies)
      .filter(([_, healthy]) => !healthy)
      .map(([dep, _]) => dep);
    
    return {
      satisfied: missing.length === 0,
      missing,
      details: dependencies
    };
  }

  startPeriodicChecks() {
    this.intervalId = setInterval(async () => {
      for (const service of this.services) {
        await this.checkServiceHealth(service);
      }
    }, this.checkInterval);
  }

  stopPeriodicChecks() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  // Mark service as initialized (for startup probe)
  markInitialized(serviceName) {
    this.initialized.add(serviceName);
  }
}

// CLI support for Docker health checks
if (require.main === module) {
  const args = process.argv.slice(2);
  if (args[0] === '--check' && args[1]) {
    const monitor = new HealthMonitor({ services: [] });
    
    switch (args[1]) {
      case 'liveness':
        monitor.livenessProbe().then(() => process.exit(0));
        break;
      case 'readiness':
        monitor.readinessProbe().then(probe => {
          process.exit(probe.ready ? 0 : 1);
        });
        break;
      case 'startup':
        monitor.startupProbe().then(probe => {
          process.exit(probe.started ? 0 : 1);
        });
        break;
      default:
        console.error('Unknown probe type:', args[1]);
        process.exit(1);
    }
  }
}

// Export the class following the interface pattern
class HealthMonitorInterface extends HealthMonitor {}

module.exports = HealthMonitorInterface;